nixos/networkd: get rid of *Config attributes in lists

Ma27 opened the pull request:
## Description of changes
This patch is about removing `wireguardPeerConfig`, `dhcpServerStaticLeaseConfig` - a.k.a. the
AbstractSingletonProxyFactoryBean of nixpkgs - and friends.

As a former colleague said

> worst abstraction ever

I second that. I've written enough networkd config for NixOS systems so far to have a strong dislike. In fact, these don't even make sense: `netdevs.wireguardPeers._.wireguardPeerConfig` will be rendered into the key `[WireGuardPeer]` and every key from `wireguardPeerConfig` is in there. Since it's INI, there's no place where sections on the same level as wireguardPeerConfig fit into. Hence, get rid of it all.

For the transition, using the old way is still allowed, but gives a warning. I think we could drop this after one release.

The tests of rosenpass and systemd-networkd-dhcpserver-static-leases were broken on the rev before, hence they were updated, but are still not building.

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

Ma27 committed:
nixos/networkd: get rid of *Config attributes in lists

This patch is about removing `wireguardPeerConfig`,
`dhcpServerStaticLeaseConfig` - a.k.a. the
AbstractSingletonProxyFactoryBean of nixpkgs - and friends.

As a former colleague said

> worst abstraction ever

I second that. I've written enough networkd config for NixOS systems so
far to have a strong dislike. In fact, these don't even make sense:
`netdevs.wireguardPeers._.wireguardPeerConfig` will be rendered into
the key `[WireGuardPeer]` and every key from `wireguardPeerConfig` is in
there. Since it's INI, there's no place where sections on the same level
as wireguardPeerConfig fit into. Hence, get rid of it all.

For the transition, using the old way is still allowed, but gives a
warning. I think we could drop this after one release.

The tests of rosenpass and systemd-networkd-dhcpserver-static-leases
were broken on the rev before, hence they were updated, but are still
not building.

WilliButz reviewed (approved):
Very nice! Also great that you managed to clean up the abstraction in a non-breaking way.

>  I've written enough networkd config for NixOS systems so far to have a strong dislike. In fact, these don't even make sense

+1 on this :)

lheckemann commented on nixos/modules/services/networking/rosenpass.nix:
```diff
@@ -130,8 +130,8 @@ in
             peer = (x: x.wireguardPeers);
-            key = (x: if x.wireguardPeerConfig ? PublicKey then x.wireguardPeerConfig.PublicKey else null);
-            description = "${options.systemd.network.netdevs}.\"<name>\".wireguardPeers.*.wireguardPeerConfig.PublicKey";
+            key = (x: if x ? PublicKey then x.PublicKey else null);
```
```suggestion
            key = (x: x.PublicKey or null);
```

Ma27 replied:
Done.

lheckemann commented on nixos/modules/system/boot/networkd.nix:
```diff
@@ -1627,34 +1627,21 @@ let
-          {manpage}`systemd.netdev(5)` for details.
-        '';
-      };
+  mkSubsectionType = oldKey: checkF:
```
(bikeshedding potential, feel free to dismiss) should this be called mkSectionType? systemd's config format has sections but no "subsections", as far as I understand.

Ma27 replied:
I called it subsection because it's one level deeper than matchConfig/networkConfig/... in the Nix abstraction.
That said, I wouldn't worry too much since I plan to rip this out with my bare hands right after the 24.11 branch-off.

lheckemann commented on nixos/modules/tasks/network-interfaces-systemd.nix:
```diff
@@ -95,65 +95,63 @@ let
-              TTLPropagate = route.options.ttl-propagate == "enabled";
-            };
-        });
+          optionalAttrs (route.address != null && route.prefixLength != null) {
```
Could we be using `mkIf` rather than chaining `optionalAttrs` here?

Ma27 replied:
Yes indeed.
Used an mkMerge/mkIf chain. Tests don't seem to have changed, so it was indeed a no-op here, but still the correct thing to do I think.

lheckemann replied:
I meant as in https://github.com/lheckemann/nixpkgs/commit/cf50bd0aa1a0d9271fd0c547f435b5c9a2a0f2c0 -- without mkMerge.

lheckemann reviewed (approved):
:heart:

Consider all my review comments as nonblocking nits.

Ma27 committed:
nixos/networking: use optionalAttrs -> mkIf for networkd route generation

Suggested in https://github.com/NixOS/nixpkgs/pull/312472#discussion_r1605894882

fpletz reviewed (approved):
:heart:

fpletz commented:
We have to wait with the merge until after the branch-off since this is a breaking change and thus unfortunately too late for 24.05.

ElvishJerricco commented:
Isn't the idea of these `*Config` options that we could have abstractions in sibling options that get rendered into the `*Config` option? Granted, I guess that hasn't actually happened yet, or else this PR wouldn't be possible.

Ma27 commented:
@ElvishJerricco for systemd.network, the structure has pretty much always been a direct mapping to how the .ini files look like, no? (Sure, with a few modifications such as using lists for sections that can appear multiple times).

I don't expect much to happen there given what the options are for. Also, I don't know anyone who likes the way this is currently structured and the usually recommended "workaround" is using "extraConfig" which is IMHO worse since it doesn't have any eval-time validation, so I'd strongly prefer my solution over the current state.

fpletz commented:
As 24.05 got branched off let's merge now and fix possible fallout of this change in the next weeks.
